crates/telio-batcher/src/batcher.rs

```diff
@@ -0,0 +1,401 @@
+use futures::future::BoxFuture;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::{collections::HashMap, sync::Arc};
+
+use telio_utils::telio_log_debug;
+use tokio::time::sleep_until;
+
+type Action<V, R = std::result::Result<(), ()>> =
+    Arc<dyn for<'a> Fn(&'a mut V) -> BoxFuture<'a, R> + Sync + Send>;
+
+pub struct Batcher<K, V> {
+    actions: HashMap<K, (BatchEntry, Action<V>)>,
+}
+
+struct BatchEntry {
+    deadline: tokio::time::Instant,
+    interval: std::time::Duration,
+    threshold: std::time::Duration,
+}
+
+impl<K, V> Default for Batcher<K, V>
+where
+    K: Eq + Hash + Send + Sync + Debug + Clone,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<K, V> Batcher<K, V>
+where
+    K: Eq + Hash + Send + Sync + Debug + Clone,
+{
+    pub fn new() -> Self {
+        Self {
+            actions: HashMap::new(),
+        }
+    }
+
+    // Batching works by sleeping until the nearest future and then trying to batch more actions
+    // based on the threshold value. Higher delay before calling the function will increase the chances of batching
+    // because the deadlines will _probably_ be in the past already.
+    pub async fn get_future(&mut self) -> Vec<(K, Action<V>)> {
+        let mut batched_actions: Vec<(K, Action<V>)> = vec![];
+
+        match self.actions.len() {
+            0 => {
+                // block if there's nothing to do. This enables the user of this function to use tokio::select properly.
+                std::future::pending::<Vec<(K, Action<V>)>>().await;
```

Comment:
With the current implementation, this code will never finish evaluating `get_future` in the assertion:
```rust
        let batcher = Arc::new(Mutex::new(Batcher::<String, TestChecker>::new()));

        let b = batcher.clone();
        let _ = tokio::spawn(async move {
            sleep(Duration::from_secs(1)).await;
            b.lock().await.add(
                "key".to_owned(),
                Duration::from_secs(1),
                Duration::from_secs(0),
                Arc::new(|s: _| {
                    Box::pin(async move {
                        s.values
                            .push(("key".to_owned(), tokio::time::Instant::now()));
                        Ok(())
                    })
                }),
            );
        });

        assert_eq!(1, batcher.lock().await.get_future().await.len());

```

Reply:
This one is deliberate. I think it makes for a nice usage in `tokio::select!` blocks. Do you have any suggestions regarding that?

I guess I could _deoptimize_ and not return a `pending()` future and just leave the call site handle everything. _should_ be safer to use. What's your opinion on this?

Reply:
What I would expect here is something like conditional variable. Sadly there is no such thing for tokio runtime. I think this can be easily simulated by:

1. wrapping match in a `loop { ... }`
2. explicit returning from the non-empty case
3. waiting for notification in empty case (eg using `tokio::sync::Notify::notified`) and letting the loop run for another iteration
4. generating a notification when new entry is added (eg using `tokio::sync::Notify::notify_waiters`)

I haven't tested this so take it with a grain of salt

Reply:
I may have misunderstood then. The comment implies that `Batcher` will never resolve, is this due to Mutex usage..?

My response about being deliberate was for returning `pending()` future when there were no items :)

Reply:
It will never resolve, because [pending](https://doc.rust-lang.org/std/future/fn.pending.html) never resolves.

Reply:
Oh okay, now I see your point. Because the task spawned will take a short-while until it will be executed and it's locked.

I've address this by returning `Future<Option<Vec<T>>>` and just using `Some(futures)` inside of the call-site. This should now work properly by not blocking unecessarily and actually looks very nice in the call site as well